airbyte-integrations/connectors/source-square/source_square/source.py

```diff
@@ -0,0 +1,419 @@
+
+import pendulum
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from source_square.utils import separate_items_by_count
+
+
+def parse_square_error_response(error):
+    if error.response.content:
+        content = json.loads(error.response.content.decode())
+        if content and "errors" in content:
+            return SquareException(error.response.status_code, content["errors"])
+
+
+class SquareStream(HttpStream, ABC):
+    def __init__(self, is_sandbox: bool, api_version: str, start_date: str, include_deleted_objects: bool, **kwargs):
+        super().__init__(**kwargs)
+        self.is_sandbox = is_sandbox
+        self.api_version = api_version
+        # Converting users ISO 8601 format (YYYY-MM-DD) to RFC 3339 (2021-06-14T13:47:56.799Z)
+        # Because this standard is used by square in 'updated_at' records field
+        self.start_date = pendulum.parse(start_date).to_rfc3339_string()
+        self.include_deleted_objects = include_deleted_objects
+
+    data_field = None
+    primary_key = "id"
+    items_per_page_limit = 100
+
+    @property
+    def url_base(self) -> str:
+        return "https://connect.squareup{}.com/v2/".format("sandbox" if self.is_sandbox else "")
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page_cursor = response.json().get("cursor", False)
+        if next_page_cursor:
+            return {"cursor": next_page_cursor}
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {"Square-Version": self.api_version, "Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        json_response = response.json()
+        records = json_response.get(self.data_field, []) if self.data_field is not None else json_response
+        yield from records
+
+    def _send_request(self, request: requests.PreparedRequest) -> requests.Response:
+        try:
+            return super()._send_request(request)
+        except requests.exceptions.HTTPError as e:
+            square_exception = parse_square_error_response(e)
+            if square_exception:
+                self.logger.error(str(square_exception))
+                exit(1)
+
+            raise e
+
+
+class SquareException(Exception):
+    """ Just for formatting the exception as Square"""
+
+    def __init__(self, status_code, errors):
+        self.status_code = status_code
+        self.errors = errors
+
+    def __str__(self):
+        return f"Code: {self.status_code}, Detail: {self.errors}"
+
+
+class SquareStreamPageParam(SquareStream, ABC):
```

Comment:
why is this one and `SquareStreampagejson` distinct classes if they do the same thing? (can you add a comment?)

Reply:
Yes, explicitly they are doing the same things - writing the next_page_token into the request.
But some streams require it in request query parameters (TeamMemberWages, Customers), but others in JSON payload (Items, Discounts, Orders, etc)

So I made 2 different classes instead of duplicating the code in each stream.
I'll add comments in the code